persistence/src/main/java/com/epam/esm/repository/impl/GiftCertificateDaoImpl.java

```diff
@@ -0,0 +1,165 @@
+    private static final String GIFT_CERTIFICATE_CREATE_DATE = "createDate";
+    private static final String TAG_NAME = "name";
+    private static final String ORDER_CERTIFICATES = "certificates";
+    private static final String ORDER_CERTIFICATES_ID = "id";
+    private static final Integer MINIMUM_SELECT_LIMIT = 1;
+
+    private final EntityManager entityManager;
+
+    @Autowired
+    public GiftCertificateDaoImpl(EntityManager entityManager) {
+        this.entityManager = entityManager;
+    }
+
+    @Override
+    public List<GiftCertificate> findAll(Integer startPosition, Integer rowsLimit) {
+        CriteriaBuilder cb = entityManager.getCriteriaBuilder();
+        CriteriaQuery<GiftCertificate> cQuery = cb.createQuery(GiftCertificate.class);
+        Root<GiftCertificate> root = cQuery.from(GiftCertificate.class);
+        return entityManager.createQuery(cQuery).setFirstResult(startPosition).setMaxResults(rowsLimit).getResultList();
+    }
+
+    @Override
+    public Optional<GiftCertificate> findById(Long id) {
+        return Optional.ofNullable(entityManager.find(GiftCertificate.class, id));
+    }
+
+    @Override
+    public GiftCertificate save(GiftCertificate giftCertificate) {
+        entityManager.persist(giftCertificate);
+        return giftCertificate;
+    }
+
+    @Override
+    public void delete(GiftCertificate certificate) {
+        entityManager.remove(certificate);
+    }
+
+    @Override
+    public GiftCertificate update(GiftCertificate giftCertificate) {
+        return entityManager.merge(giftCertificate);
+    }
+
+    @Override
+    public List<GiftCertificate> search(GiftCertificateSearchParams searchParams,
+                                        Integer startPosition,
+                                        Integer rowsLimit) {
+        CriteriaBuilder cb = entityManager.getCriteriaBuilder();
+        CriteriaQuery<GiftCertificate> cQuery = cb.createQuery(GiftCertificate.class);
+        Root<GiftCertificate> root = cQuery.from(GiftCertificate.class);
+
+        List<Predicate> searchByParameters = setSearchParameters(searchParams, cb, root);
+
+        cQuery.select(root).where(cb.and(searchByParameters.toArray(new Predicate[searchByParameters.size()])));
+        setOrderParameters(cb, cQuery, root, searchParams);
+
+        return entityManager.createQuery(cQuery).setFirstResult(startPosition).setMaxResults(rowsLimit).getResultList();
+    }
+
+    @Override
+    public boolean isGiftCertificateUsedInOrder(Long id) {
+        CriteriaBuilder cb = entityManager.getCriteriaBuilder();
+        CriteriaQuery<Order> cQuery = cb.createQuery(Order.class);
+        Root<Order> root = cQuery.from(Order.class);
+        cQuery.select(root).where(cb.equal(root.join(ORDER_CERTIFICATES).get(ORDER_CERTIFICATES_ID), id));
+        List<Order> orders = entityManager.createQuery(cQuery).setMaxResults(MINIMUM_SELECT_LIMIT).getResultList();
+        return orders.isEmpty() || orders == null ? false : true;
```

Comment:
looks like NPE in case if orders==null

NPE is 80% of java exceptions

Reply:
fixed